src/gui/dialogs/DialogStateful.cpp

```diff
@@ -0,0 +1,131 @@
+#include "DialogStateful.hpp"
+#include "DialogRadioButton.hpp"
+#include "gui.h"
+
+extern window_t *window_1; //current popup window, C-code remain
+
+//*****************************************************************************
+
+IDialogStateful::IDialogStateful(const char *name, int16_t WINDOW_CLS_)
+    : IDialog(WINDOW_CLS_)
+    , id_capture(window_capture())
+    , color_back(gui_defaults.color_back)
+    , color_text(gui_defaults.color_text)
+    , font(gui_defaults.font)
+    , font_title(gui_defaults.font_big)
+    , padding(gui_defaults.padding)
+    , flags(0)
+    , last_text_h(0)
+    , phase(0)
+    , progress(-1)
+    , title(name) {
+    window_1 = this; //todo
+    gui_reset_jogwheel();
+    gui_invalidate();
+    window_set_capture(id);
+}
+
+bool IDialogStateful::Change(uint8_t phs, uint8_t progress_tot, uint8_t progr) {
+    if (!can_change(phs))
+        return false;
+    if (phase != phs) {
+        phase = phs;
+        flags |= DLG_PHA_CH;
+    }
+    if (progress_tot != progress) {
+        progress = progress_tot;
+        flags |= DLG_PRO_CH;
+    }
+    gui_invalidate();
+    return true;
+}
+
+IDialogStateful::~IDialogStateful() {
+    window_destroy(id);
+    window_set_capture(id_capture);
+    window_invalidate(0);
+}
+
+extern window_t *window_1; //current popup window
+
+void IDialogStateful::draw_frame() {
+    rect_ui16_t rc = rect;
+    display->draw_line(point_ui16(rc.x, rc.y), point_ui16(239, rc.y), COLOR_GRAY);
+    display->draw_line(point_ui16(rc.x, rc.y), point_ui16(rc.x, 320 - 67), COLOR_GRAY);
+    display->draw_line(point_ui16(239, rc.y), point_ui16(239, 320 - 67), COLOR_GRAY);
+    display->draw_line(point_ui16(rc.x, 320 - 67), point_ui16(239, 320 - 67), COLOR_GRAY);
+}
+
+//this should be moved elswhere
```

Comment:
Use TODO in such comment, it can be find easily.

Reply:
fixed